Add "Common Functions" as per spec, including unit tests

PeterTh opened the pull request:
Also refactors testing and extends test coverage to vec8 and vec16.

Note that the functions are mostly implemented in a straightforward fashion. There might be more efficient mappings of some of them to some backends (e.g. CUDA), but I think for now it's most important to be standard compliant.

illuhad commented on include/CL/sycl/common_functions.hpp:
```diff
@@ -0,0 +1,140 @@
+
+namespace detail {
+  template<class T>
+  constexpr T pi = T(3.1415926535897932385L);
```
In `math.hpp` we use `M_PI` - we should probably be consistent here and either use `detail::pi` everywhere or use `M_PI` everywhere. I'm fine with both :)

illuhad commented on include/CL/sycl/common_functions.hpp:
```diff
@@ -0,0 +1,140 @@
+}
+
+template<typename T>
+inline auto clamp(T x, T minval, T maxval) {
```
please mark all functions intended to be used inside kernels with `HIPSYCL_KERNEL_TARGET` (expands to `__host__ __device__` currently). This is required for the support of "manual mode" and "legacy source-to-source transformation", although the latter is deprecated. See [here](https://github.com/illuhad/hipSYCL/issues/42#issuecomment-483312569) for more details.

PeterTh commented:
Done. 
I decided to just use M_PI (even though it's not-100%-standard) since in practice it works everywhere.

illuhad commented:
Thanks! I'm still getting
```
unit_tests.cpp:1291:24: error: call of overloaded ‘min(cl::sycl::vec<float, 8>&, cl::sycl::vec<float, 8>&)’ is ambiguous
```
(and lots of the same, and same for `max`). This is because your new `min()` and `max()` using `fmin` and `fmax` collide with the ones from `builtin.hpp` for integers. So, we should probably only enable the `fmin` and `fmax` variants for floating point arguments.

PeterTh commented:
To repro and fix this I added a unit test for integer min/max. 

However, even if I limit my implementations to `genfloat` I'm facing the problem that the min/max in `builtin.hpp` also applies to `float` -- should this be the case?

On the other hand, if I remove the ambiguity by removing my versions of `min(T,T)` and `max(T,T)` entirely, I run into the issue that the version provided in `builtin.hpp`, while entering the overload set and causing ambiguity, doesn't actually correctly instantiate for all the cases I test.

I don't fully understand the intention behind the definitions/macros in `builtin.hpp`, so I'm not sure how to proceed. Should I rewrite builtin.hpp to only provide the `geninteger` versions?

illuhad commented:
You're right, there's a bug! `builtin.hpp` was indeed intended to be for integer builtins (contrary to `math.hpp` for floats), but the `HIPSYCL_DEFINE_BUILTIN` macro doesn't correctly limit itself to integer types. Additionally, that header seems to only have partially made the transition to the new clang plugin where the `HIPSYCL_KERNEL_TARGET` macro was used. `builtin.hpp` still uses straight `__device__` attributes.

So, I see two ways to proceed:
* You fix the `builtin.hpp` header to only accept integer types
* Or you comment out the `min/max` unit tests and I fix the `builtin.hpp` header, and also bring it up to speed with the current clang plugin setup. I would then reenable the unit tests once `builtin.hpp` is fixed.

PeterTh commented:
I've resolved the ambiguities and added some basic testing for `geninteger` min/max.  Note that the existing implementation broke when I added `char` (not unsigned), and vectors thereof, to the list of types tested -- for now I just skipped that.

illuhad commented:
Excellent, thank you!
